src/util/apparmor.c

```diff
@@ -93,3 +105,446 @@ int bus_apparmor_dbus_supported(bool *supportedp) {
+ * old labels. In this case we treat this as if the ownership request was
+ * denied.
+ *
+ * Return: 0 if the ownership is allowed, BUS_APPARMOR_E_DENIED if it is not,
+ *         or a negative error code on failure.
+ */
+int bus_apparmor_check_own(struct BusAppArmorRegistry *registry,
+                           const char *owner_context,
+                           const char *name) {
+        _c_cleanup_(c_freep) char *condup = NULL;
+        char *security_label, *security_mode;
+        _c_cleanup_(c_freep) char *qstr = NULL;
+        int r;
+        /* the AppArmor API uses pointers to int for pointers to boolean */
+        int allow = false, audit = true;
+
+        if (!is_apparmor_enabled(registry))
+                return 0;
+
+        condup = strdup(owner_context);
+        if (!condup)
+                return error_origin(-ENOMEM);
+
+        security_label = aa_splitcon(condup, &security_mode);
+
+        r = build_service_query(&qstr, security_label, registry->bustype, name);
+        if (r < 0)
+                return error_origin(r);
+
+        r = aa_query_label(AA_DBUS_BIND, qstr, r, &allow, &audit);
+        if (r < 0)
+                return error_origin(-errno);
+
+        if (audit)
+                bus_apparmor_log("apparmor=\"%s\" operation=\"dbus_bind\" bus=\"%s\" name=\"%s\"", allowstr(allow), registry->bustype, name);
+
+        if (is_complain(security_mode))
+                allow = 1;
+
+        if (!allow)
+                return BUS_APPARMOR_E_DENIED;
+
+        return 0;
+}
+
+/**
+ * bus_apparmor_check_xmit() - check if the given transaction is allowed
+ * @registry:           AppArmor registry to operate on
+ * @check_send:         true if sending should be checked, false is receiving should be checked
+ * @sender_context:     security context of the sender
+ * @receiver_context:   security context of the receiver, or NULL
+ * @subject:            List of names
+ * @subject_id:         Unique ID of the subject
+ * @path:               Dbus object path
+ * @interface:          DBus method interface
+ * @method:             DBus method that is being called
+ *
+ * Check if the given sender context is allowed to send/receive a message
+ * to the given receiver context. If the any context is given as NULL,
+ * the per-registry fallback context is used instead.
+ *
+ * The contexts are pinned when peers connect, and as such could in principle
+ * become invalid in case a new policy is loaded that does not know the
+ * old labels. In this case we treat this as if the transaction was
+ * denied.
+ *
+ * In case multiple names are available all are being checked and the function
+ * will deny access if any of them is denied by AppArmor.
+ *
+ * Return: 0 if the transaction is allowed, BUS_APPARMOR_E_DENIED if it is not,
+ *         or a negative error code on failure.
+ */
+int bus_apparmor_check_xmit(BusAppArmorRegistry *registry,
+                            bool check_send,
```

Comment:
dbus-daemon only ever checks on `send`, why do we have two calls one for send and one for receive?

Maybe this is a confusion where dbus-daemon checks both for sending and receival in a single call, but that is not the case. Checking at send and receive is different to checking the receiver during send. I think this should be `bus_apparmor_check_send()` just like selinux does.